keras/layers/recurrent.py

```diff
@@ -78,11 +128,27 @@ def build(self):
         self.U = self.inner_init((self.output_dim, self.output_dim))
         self.b = shared_zeros((self.output_dim))
         self.params = [self.W, self.U, self.b]
-
+        nw = len(self.initial_weights) if self.initial_weights is not None else 0
+        if stateful:
+            if self.initial_state is not None:
+                self.h = sharedX(self.initial_state[0])
+                del self.initial_state
+            elif self.batch_size is not None:
+                self.h = shared_zeros((self.batch_size, self.output_dim))
+            elif self.initial_weights is not None:
+                if nw == len(self.params) + 1:
+                    self.h = sharedX(self.initial_weights[-1])
+                    nw -= 1
+                else:
+                    raise Exception("Hidden state not provided in weights")
+            else:
+                raise Exception("One of the following arguments must be provided for stateful RNNs: hidden_state, batch_size, weights")
+        self.state = [self.h]
         if self.initial_weights is not None:
-            self.set_weights(self.initial_weights)
+            self.set_weights(self.initial_weights[:nw])
             del self.initial_weights
 
+
```

Comment:
the next _step function uses `return self.activation(x_t + mask_tm1 * T.dot(h_tm1, u))` which makes statefulness completly useless when using different length sequences. It should instead return
```python
h_t = self.activation( x_t + T.dot(h_tm1, u))
return mask * h_t + (1 - mask) * h_tm1
```
I actually have a question. What is faster inside a scan loop?
* `mask * h_t + (1 - mask) * h_tm1` or
*  `T.switch(T.eq(mask, 1), h_t, h_tm1)`

@nouiz, if you have time, could you please say your opinion? Will both have the same gradient optimized gradient?